### ISSUE TITLE
Unhandled rejection Error: listen EACCES 0.0.0.0:443

### ISSUE THREAD

**Issue by xingped** [Daplie/letsencrypt-express#91]
I'm trying to use Apache for my server in order to have multiple domains on a single server, but using Passenger to execute the node process. It works all fine and dandy for plain express and normal Port 80 stuff, but when I try to use your module for SSL it throws a hissy fit.

Any idea what might be wrong?

Error log below:

    App 18674 stdout: 
    App 18674 stderr: (node:18674) DeprecationWarning: 'GLOBAL' is deprecated, use 'global'
    App 18674 stdout: Handling ACME challenges and redirecting to https on plain port 80
    App 18674 stderr: Unhandled rejection Error: listen EACCES 0.0.0.0:443
    App 18674 stderr:     at Object.exports._errnoException (util.js:1036:11)
    App 18674 stderr:     at exports._exceptionWithHostPort (util.js:1059:20)
    App 18674 stderr:     at Server._listen2 (net.js:1239:19)
    App 18674 stderr:     at listen (net.js:1288:10)
    App 18674 stderr:     at Server.listen (net.js:1384:5)
    App 18674 stderr:     at /home/ec2-user/vhosts/mysite.com/node_modules/letsencrypt-express/lex.js:40:92
    App 18674 stderr:     at Promise._execute (/home/ec2-user/vhosts/mysite.com/node_modules/bluebird/js/release/debuggability.js:299:9)
    App 18674 stderr:     at Promise._resolveFromExecutor (/home/ec2-user/vhosts/mysite.com/node_modules/bluebird/js/release/promise.js:481:18)
    App 18674 stderr:     at new Promise (/home/ec2-user/vhosts/mysite.com/node_modules/bluebird/js/release/promise.js:77:14)
    App 18674 stderr:     at /home/ec2-user/vhosts/mysite.com/node_modules/letsencrypt-express/lex.js:39:21
    App 18674 stderr:     at Array.forEach (native)
    App 18674 stderr:     at Object.opts.listen (/home/ec2-user/vhosts/mysite.com/node_modules/letsencrypt-express/lex.js:38:11)
    App 18674 stderr:     at Object.<anonymous> (/home/ec2-user/vhosts/mysite.com/index.js:15:5)
    App 18674 stderr:     at Module._compile (module.js:556:32)
    App 18674 stderr:     at Object.Module._extensions..js (module.js:565:10)
    App 18674 stderr:     at Module.load (module.js:473:32)

My simple index.js file:

    var express = require('express');
    var app = express();
    var lex = require('letsencrypt-express').create({
      server: 'staging',
      email: 'myemail@email.com',
      agreeTos: true,
      approveDomains: ['mysite.com', 'www.mysite.com'],
      app: app
    });

    app.use(express.static(__dirname + '/public'));

    lex.listen(80, 443);

**Comment by Krail** [comment]
Is another instance running? Or is another app using port 443?

What's the output of `netstat -a` and `ps aux | grep node`?

**Comment by xingped** [comment]
`netstat -a | grep node` returns the following:

    unix  2      [ ACC ]     STREAM     LISTENING     99889652 /tmp/passenger.8eC9Pdf/apps.s/node.vlxjc


`ps aux | grep node` returns nothing.

It's my understanding that Passenger should be handling these port bindings, but I'm not sure how that may interfere with this library.

**Comment by Krail** [comment]
Do `netstat -a`, not `netstat -a | grep node`. The point of the first one is to see if another app is using the port 443. Also, try running the app as superuser.

**Comment by xingped** [comment]
Ah, I see. `sudo netstat -a` returns the following:

    Active Internet connections (servers and established)
    Proto Recv-Q Send-Q Local Address               Foreign Address             State      
    tcp        0      0 *:ssh                       *:*                         LISTEN      
    tcp        0      0 *:47703                     *:*                         LISTEN      
    tcp        0      0 localhost:smtp              *:*                         LISTEN      
    tcp        0      0 *:sunrpc                    *:*                         LISTEN      
    tcp        0    252 ip-172-31-25-150.us-wes:ssh c-73-109-110-168.hsd1:41062 ESTABLISHED 
    tcp        0      0 *:http                      *:*                         LISTEN      
    tcp        0      0 *:ssh                       *:*                         LISTEN      
    tcp        0      0 *:https                     *:*                         LISTEN      
    tcp        0      0 *:48383                     *:*                         LISTEN      
    tcp        0      0 *:sunrpc                    *:*                         LISTEN      
    udp        0      0 *:sunrpc                    *:*                                     
    udp        0      0 *:asia                      *:*                                     
    udp        0      0 ip-172-31-25-150.us-west:ntp *:*                                     
    udp        0      0 localhost:ntp               *:*                                     
    udp        0      0 *:ntp                       *:*                                     
    udp        0      0 localhost:659               *:*                                     
    udp        0      0 *:55193                     *:*                                     
    udp        0      0 *:bootpc                    *:*                                     
    udp        0      0 *:sunrpc                    *:*                                     
    udp        0      0 *:asia                      *:*                                     
    udp        0      0 *:34473                     *:*                                     
    Active UNIX domain sockets (servers and established)
    Proto RefCnt Flags       Type       State         I-Node Path
    unix  2      [ ACC ]     STREAM     LISTENING     99889399 /tmp/passenger.8eC9Pdf/agents.s/watchdog_api
    unix  2      [ ACC ]     STREAM     LISTENING     99889443 /tmp/passenger.8eC9Pdf/agents.s/core
    unix  2      [ ACC ]     STREAM     LISTENING     99889444 /tmp/passenger.8eC9Pdf/agents.s/core_api
    unix  2      [ ACC ]     STREAM     LISTENING     99889491 /tmp/passenger.8eC9Pdf/agents.s/ust_router
    unix  2      [ ACC ]     STREAM     LISTENING     99889492 /tmp/passenger.8eC9Pdf/agents.s/ust_router_api
    unix  2      [ ACC ]     STREAM     LISTENING     99889652 /tmp/passenger.8eC9Pdf/apps.s/node.vlxjc
    unix  2      [ ACC ]     STREAM     LISTENING     7897   @/com/ubuntu/upstart
    unix  2      [ ACC ]     SEQPACKET  LISTENING     8162   @/org/kernel/udev/udevd
    unix  11     [ ]         DGRAM                    9105   /dev/log
    unix  2      [ ACC ]     STREAM     LISTENING     9240   /var/run/rpcbind.sock
    unix  2      [ ACC ]     STREAM     LISTENING     680749 /var/run/dbus/system_bus_socket
    unix  3      [ ]         STREAM     CONNECTED     99889449 
    unix  3      [ ]         DGRAM                    8171   
    unix  3      [ ]         STREAM     CONNECTED     99889578 
    unix  2      [ ]         DGRAM                    99938540 
    unix  3      [ ]         STREAM     CONNECTED     680751 
    unix  3      [ ]         STREAM     CONNECTED     99889579 
    unix  2      [ ]         DGRAM                    680713 
    unix  2      [ ]         DGRAM                    680687 
    unix  3      [ ]         STREAM     CONNECTED     99889402 
    unix  2      [ ]         DGRAM                    99938903 
    unix  2      [ ]         DGRAM                    680741 
    unix  3      [ ]         STREAM     CONNECTED     99938544 
    unix  2      [ ]         DGRAM                    826131 
    unix  3      [ ]         DGRAM                    8170   
    unix  3      [ ]         STREAM     CONNECTED     99889367 
    unix  2      [ ]         DGRAM                    10968  
    unix  3      [ ]         STREAM     CONNECTED     99889448 
    unix  3      [ ]         STREAM     CONNECTED     99938545 
    unix  3      [ ]         STREAM     CONNECTED     680752 
    unix  3      [ ]         STREAM     CONNECTED     99889368 
    unix  2      [ ]         DGRAM                    9316   
    unix  3      [ ]         STREAM     CONNECTED     99889403 
    unix  2      [ ]         DGRAM                    680729 
    unix  2      [ ]         DGRAM                    603232

**Comment by xingped** [comment]
And my vhosts.conf file (if it helps): 

    <VirtualHost *:80>
      ServerName mysite.com
      ServerAlias www.mysite.com
      Redirect / https://mysite.com/
    </VirtualHost>
    
    <VirtualHost *:443>
      ServerName mysite.com
      ServerAlias www.mysite.com
    
      # Tell Apache and Passenger where your app's code directory is
      DocumentRoot /var/www/vhosts/mysite.com
      PassengerAppRoot /var/www/vhosts/mysite.com
    
      # Tell Passenger that your app is a Node.js app
      PassengerAppType node
      PassengerStartupFile index.js
    
    #  SSLEngine on
    #  SSLCertificateFile /etc/letsencrypt/live/mysite.com/cert.pem
    #  SSLCertificateKeyFile /etc/letsencrypt/live/mysite.com/privkey.pem
    #  SSLCertificateChainFile /etc/letsencrypt/live/mysite.com/chain.pem
    #  SSLProtocol All -SSLv2 -SSLv3
    #  SSLHonorCipherOrder on
    #  SSLCipherSuite "EECDH+ECDSA+AESGCM EECDH+aRSA+AESGCM EECDH+ECDSA+SHA384 EECDH+ECDSA+SHA256 EECDH+aRSA+SHA384 EECDH+aRSA+SHA256 EECDH+aRSA+RC4 EECDH EDH+aRSA !RC4 !aNULL !eNULL !LOW !3DES !MD5 !EXP !PSK !SRP !DSS"
    
      # Relax Apache security settings
      <Directory "/var/www/vhosts/mysite.com">
        Require all granted
        Options +FollowSymLinks
        AllowOverride None
        Order allow,deny
        Allow from all
        Options -MultiViews
        Satisfy Any
      </Directory>
    </VirtualHost>

I suppose it may or may not have something to do with my SSL stuff being currently commented out (because for whatever reason apache can't seem to access my key files), but that seems unrelated to being unable to bind to a port, doesn't it?

**Comment by Krail** [comment]
1. Are you running your server as root/superuser? [source](http://stackoverflow.com/a/18947576)
2. I think this line may be indicating that another process is currently utilizing port 443, but I'm not sure.
```
tcp        0      0 *:https                     *:*                         LISTEN
```

**Comment by xingped** [comment]
1. Yes, I am. Node applications on Port 80 run fine that use app.listen(80) to bind to the port.

2. It's probably Apache, but I'm not sure how to verify? :http is also running and node apps on port 80 run fine.

**Comment by Krail** [comment]
I have zero experience running nodejs on an Apache HTTP server. You're going to need to research on your own. I found [this](http://stackoverflow.com/a/18604082) which may or may not help.

Anyways, this is **not** an issue with this project and should be closed.

**Comment by insightfuls** [comment]
Disclaimer: I have no actual experience with Passenger, but here's my take on it.

This won't be working because Passenger only supports `http` servers in Node, not `https` ones. It uses a reverse port binding mechanism to achieve this. Passenger/Apache will listen on port 80, and will force Node to listen on another port, and Passenger redirects the requests appropriately. There are problems if you want Node to listen on multiple ports. See https://www.phusionpassenger.com/library/indepth/nodejs/reverse_port_binding.html

In the case where you want to add SSL, Passenger/Apache add the security layer and listen on port 443, but that's the only difference. They still expect Node to serve plain HTTP content on an overridden port. Most importantly, the TLS negotiation is performed by Apache (before the underlying HTTP request can even be made), so it's Apache, not Node, that needs to have the certificates installed. I notice you commented out the SSL stuff in your Apache configuration. That will turn off the SSL security and will result in port 443 responding to plain HTTP, not HTTPS; it won't simply pass the responsibility for handling the security layer to Node. When used with Apache, Apache interprets the HTTP request and forwards only appropriate ones on to Passenger to be handled by Node. You can't move the SSL wrapping to Node, making the HTTP inaccessible to Apache, because then Apache couldn't play its role in the whole setup. It's not just port forwarding.

The bottom line of all this is that `letsencrypt-express` can't fit into this setup.

If you wish to use Apache to handle SSL (and if you're using Passenger with Apache, Apache has to handle the SSL) you must configure your certificates in Apache. You could do this using `letsencrypt-cli` with the webroot option (modify your Apache config to serve /.well-known/acme-challenge/ on port 80 rather than redirecting that to HTTPS port 443), or my forthcoming apache/hooks support which can run entirely on port 443. The official Certbot client could also be used both these ways. All of these will require a systemd/cron job to automate renewal.

Hope this helps.

**Comment by xingped** [comment]
@insightfuls That was extremely helpful, thank you! I didn't really understand how Passenger worked. And I definitely didn't know it couldn't handle https. What is the upcoming apache/hooks support you're referencing?

I'll close the issue then since it's not doable the way I wanted.

**Comment by insightfuls** [comment]
The upcoming Apache/hooks option is documented in the repository here (see TLS SNI, production option 2): https://github.com/Daplie/letsencrypt-cli/blob/master/README.md

I'm reworking the options etc. though to be a bit more generic, as commented in this PR: https://github.com/Daplie/letsencrypt-cli/pull/24

So when it gets into an actual release it will look a bit different. That could be a few days away, but won't be too long.

**Comment by xingped** [comment]
@insightfuls I don't think this registered the first time I read your reply, but it just hit me and I just wanted to confirm - you're saying that I can let Apache handle all the encryption and the traffic between Node and Apache will be plaintext, but encrypted between Apache and the client?

**Comment by insightfuls** [comment]
Yes, that's right. That's how Passenger would work with Apache. And therefore it's Apache that needs to be configured for SSL, and not Node.

**Comment by xingped** [comment]
Awesome, thanks!

**Comment by joshuakcockrell** [comment]
@ghost Running node servers as root/superuser is a big security risk. That should not be the solution here.
